Logarithmic timestamp comparison for downscaling

damemi opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Implements logarithmically-compared timestamps for replica scale downs, from https://github.com/kubernetes/kubernetes/pull/96898:

> Compares ready and creation timestamps in a logarithmic scale. This allows for some level of randomness when Pods are quick-sorted to get downscaling candidates.
>
> Used base 2. This means that (roughly) if a Pod A has been created/running for less than half the time of Pod B, then Pod A will be downscaled first. But if Pod A has been created/running for more than half the time of Pod B, they can be equally downscaled.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref https://github.com/kubernetes/enhancements/issues/2185 and https://github.com/kubernetes/kubernetes/issues/96748

#### Special notes for your reviewer:
> This is a proposal that has very low overhead compared to #96748. Since behavior is not backwards compatible, we could release with a FeatureGate first. Then, from feedback, we can adjust the logarithmic base or, if we find out that the behavior might not be desired by everyone, we could make it a configuration option.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
When downscaling ReplicaSets, ready and creation timestamps are compared in a logarithmic scale.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

alculquicondor committed:
Logarithmic timestamp comparison for ReplicSet downscaling

Change-Id: I0657ea0ce41b98fdee1a5307b5826a10deaff98c

damemi reviewed (commented):
/cc @alculquicondor

damemi commented:
/retest

alculquicondor commented:
You forgot to add the KEP to the description :)

alculquicondor commented on pkg/controller/controller_utils.go:
```diff
@@ -808,6 +809,9 @@ func (s ActivePods) Less(i, j int) bool {
+// In 5 and 7, times are compared in a logarithmic scale. This allows a level
+// of randomness among equivalent Pods when sorting.
+//
 // If none of these rules matches, the second pod comes before the first pod.
```
What about the UUID comparison that was suggested in the KEP?

alculquicondor commented on test/integration/replicationcontroller/replicationcontroller_test.go:
```diff
@@ -494,6 +495,45 @@ func TestSpecReplicasChange(t *testing.T) {
 	}
 }
 
+func TestLogarithmicScaleDown(t *testing.T) {
```
This test is ensuring that the existing behavior somewhat still holds. That is good. We should run it with the feature gate enabled and disabled (I would remove the `Logarithmic` part from the test name)

Another test we could have is rather opposite: have all the pods run at roughly the same time, downscale, and see if some level of randomness holds. Do you think something like this is possible? Maybe we can run the same test X times and ensure that at least in one of them, the removed pod was not the absolute youngest.

In the test plan we also suggested emulating the story https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2185-random-pod-select-on-replicaset-downscale#test-plan
We can do that in a follow up.

damemi replied:
> have all the pods run at roughly the same time, downscale, and see if some level of randomness holds ... Maybe we can run the same test X times

There is an inherent problem with randomness in that we can never be certain it will be the randomness we expect :) Something like this will introduce flakes, which can be minimized by increasing the number of test runs, but still always possible. Are there examples of other tests which take a similar approach?

What we really wish to show from such a test is that pods created on a similar logarithmic time scale have an equal chance of being chosen for downscaling. In other words, that their ranks are calculated properly. I believe the unit tests cover this level of detail sufficiently. 

The integration test is simply showing that the basic logic still works when removed from the vacuum of unit tests. I think that the user story you linked will make a good e2e, and I could actually add that to this PR (I don't see any reason to wait, and that will flesh out test cases here more comfortably). Wdyt?

alculquicondor replied:
> I believe the unit tests cover this level of detail sufficiently.

I agree, but thanks for giving it a thought.

Still, modify the test to run for feature gate enabled and disabled.

Fine with me to add the 2e2 test for the user story in this PR.

alculquicondor reviewed (commented):
We are missing the feature gate

damemi commented:
@alculquicondor I added the UID sorting right before the call to actually sort the pods in 5ffacd0a98bb855bcbe067b2560a6f56c8775254. Let me know if that looks good, it should at least give a pseudo-random start before pods of similar ranks are sorted.

I am still working on how to add the e2e. I'm thinking these steps:
1. Create `X` pods
2. Wait 30 seconds (this should be enough that those pods will have a rank of `4`)
3. Create `Y` more pods
4. Wait another 30 seconds (now the first pods should be ~60s old w/rank `6`, and the new pods have rank `4`)
5. Downscale by `Y` pods
6. Confirm that the remaining pods are all from the original group `X`

My goal is to create 2 groups of pods, spaced far enough apart that they will have different ranks. However I add the 30s waits to reduce flakes caused by latency (in other words, the difference between a base-2 rank `2` and rank `3` is only 4 seconds, but rank `5` and `6` is 32 seconds).

What do you think of this approach? I want to avoid adding too much "scheduling" dependency into the test (ie, assuming that pods will be evenly distributed among nodes) because that introduces flakes unless the nodes are evenly balanced and I think it's irrelevant to this anyway.

alculquicondor commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -802,6 +802,10 @@ func getPodsToDelete(filteredPods, relatedPods []*v1.Pod, diff int) []*v1.Pod {
 	if diff < len(filteredPods) {
 		podsWithRanks := getPodsRankedByRelatedPodsOnSameNode(filteredPods, relatedPods)
+		// First obtain a pseudo-random ordering by sorting by pod UID
+		sort.Slice(podsWithRanks.Pods, func(i, j int) bool {
```
Why not just add the comparison in the existing `Less`? I feel like that should perform better.

damemi replied:
Should that come before or after the timestamp comparison?

alculquicondor replied:
it's a comparison criteria with the lowest priority, so after.

damemi replied:
in that case I should also update [`logarithmicOrAfterZero`](https://github.com/kubernetes/kubernetes/pull/99212/files#diff-8128d56d17ba6f10e337cf42856a93fcfaf5bce5b7efbc6d9a371bb5c04c7098R908) to check if the 2 ranks are equal (and continue), rather than just returning... or I can add the UID check right into `logarithmicOrAfterZero`. Wdyt?

alculquicondor replied:
probably better to be explicit about UIDs by having that comparison outside of timestamp comparison.

alculquicondor commented:
Re: E2E

I think we should do what was suggested in the KEP: simulate the scenario from the story and make sure spreading is good, with some allowance for skew.

alculquicondor commented on pkg/controller/controller_utils.go:
```diff
@@ -929,13 +925,21 @@ func afterOrZero(t1, t2 *metav1.Time) bool {
 // logarithmicRank calculates the base-2 logarithmic ranks of 2 timestamps,
 // compared to the current timestamp
-func logarithmicRank(t1, t2, now metav1.Time) (int64, int64) {
+func logarithmicRank(t1, t2, now metav1.Time) int64 {
```
nit: logarithmicRankDiff

alculquicondor commented on pkg/controller/controller_utils.go:
```diff
@@ -904,16 +902,14 @@ func (s ActivePodsWithRanks) Less(i, j int) bool {
+		if rankDiff > 0 {
+			return true
+		}
+		return s.Pods[i].UID < s.Pods[j].UID
```
can you add this to the comment for `ActivePodsWithRanks`?

alculquicondor reviewed (commented):
good to squash

alculquicondor commented on pkg/controller/controller_utils.go:
```diff
@@ -807,6 +808,10 @@ func (s ActivePods) Less(i, j int) bool {
 //
+// In 5 and 8, times are compared in a logarithmic scale. This allows a level
+// of randomness among equivalent Pods when sorting. If two pods have the same
+// logarithmic rank, they are sorted by UUID.
```
to provide a pseudorandom order

alculquicondor commented on pkg/controller/controller_utils.go:
```diff
@@ -872,7 +881,18 @@ func (s ActivePodsWithRanks) Less(i, j int) bool {
+				return afterOrZero(readyTime1, readyTime2)
+			}
+			rankDiff := logarithmicRankDiff(*readyTime1, *readyTime2, s.Now)
+			fmt.Printf("RANKDIFF: %+v, podi: %+v, podj: %+v\n\n", rankDiff, s.Pods[i].Name, s.Pods[j].Name)
```
leftover :)

damemi replied:
whoops! debugging

alculquicondor reviewed (commented):
good for squash

alculquicondor commented:
/assign @kow3ns

soltysh commented on pkg/controller/controller_utils.go:
```diff
@@ -807,6 +808,10 @@ func (s ActivePods) Less(i, j int) bool {
 // 8. If the pods' creation times differ, the pod that was created more recently
 //    comes before the older pod.
 //
+// In 5 and 8, times are compared in a logarithmic scale. This allows a level
```
```suggestion
// In 6 and 8, times are compared in a logarithmic scale. This allows a level
```

soltysh commented on pkg/controller/controller_utils.go:
```diff
@@ -881,7 +900,17 @@ func (s ActivePodsWithRanks) Less(i, j int) bool {
+		if rankDiff > 0 {
+			return false
+		}
+		return s.Pods[i].UID < s.Pods[j].UID
```
The logic is identical for both pieces, how about making this a function rather than repeating the code?

alculquicondor replied:
The idea is to keep the UID comparison in this method, for visibility. But a shorter alternative would be

```golang
diff := logarithmicRankDiff(s.Pods[i].CreationTimestamp, s.Pods[j].CreationTimestamp, s.Now)
if diff == 0 {
  return s.Pods[i].UID < s.Pods[j].UID
}
return diff < 0
```

damemi replied:
Yeah, I did it my way to keep the UID comparison physically last since it has the lowest priority. But if this is cleaner, we can do that too

soltysh commented on pkg/controller/controller_utils.go:
```diff
@@ -895,6 +924,25 @@ func afterOrZero(t1, t2 *metav1.Time) bool {
+		r2 = -1
+	} else {
+		r2 = int64(math.Log2(float64(d2)))
+	}
```
```golang
r1 := int64(-1)
r2 := int64(-1)
if d1 > 0 {
	r1 = int64(math.Log2(float64(d1)))
}
if d2 > 0 {
	r2 = int64(math.Log2(float64(d2)))
}
```
Seems simpler, no?

alculquicondor replied:
+1

soltysh reviewed (approved):
/approve
Some minor nits.

damemi commented:
Thanks for the review @soltysh. I pushed and rebased with the feedback, and also added the feature gate @alculquicondor pointed out from https://github.com/kubernetes/kubernetes/pull/99212#pullrequestreview-593731060. Please let me know if there is anything else I need to do

alculquicondor reviewed (commented):
You need to enable the feature gate for the tests.

I had long forgotten about the gate 😂

soltysh reviewed (approved):
/lgtm
/triage accepted
/priority backlog
